ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/methods/AdminNodeInfo.java

```diff
@@ -112,4 +126,39 @@ public JsonRpcResponse response(final JsonRpcRequestContext requestContext) {
 
     return new JsonRpcSuccessResponse(requestContext.getRequest().getId(), response);
   }
+
+  private URI getNodeAsString(
+      final EnodeURL enodeURL, final String ip, final int listeningPort, final int discoveryPort) {
+    final String uri =
+        String.format(
+            "enode://%s@%s:%d", enodeURL.getNodeId().toUnprefixedString(), ip, listeningPort);
+    if (listeningPort != discoveryPort) {
+      return URI.create(uri + String.format("?discport=%d", discoveryPort));
+    } else {
+      return URI.create(uri);
+    }
+  }
+
+  private String getIp(final EnodeURL enode) {
+    return natService
+        .filter(NatService::isNatEnvironment)
+        .flatMap(NatService::queryExternalIPAddress)
+        .orElseGet(enode::getIpAsString);
+  }
+
+  private int getDiscoveryPort(final EnodeURL enode) {
+    return natService
+        .filter(NatService::isNatEnvironment)
+        .flatMap(s -> s.getPortMapping(NatServiceType.DISCOVERY, NetworkProtocol.UDP))
+        .map(NatPortMapping::getExternalPort)
+        .orElseGet(enode::getDiscoveryPortOrZero);
```

Comment:
What does it mean if we return 0 here?

Reply:
we return 0 when `isListening` or `isRunningDiscovery` return false but the port is only used : 

- when `isListening` and` isRunningDiscovery` is true
- in the `getNodeAsString` in order to display discport when `listeningPort != discoveryPort`
```
public boolean isRunningDiscovery() {
    return discoveryPort.isPresent();
  }

  public OptionalInt getListeningPort() {
    return listeningPort;
  }

  public int getListeningPortOrZero() {
    return listeningPort.orElse(0);
  }

  public OptionalInt getDiscoveryPort() {
    return discoveryPort;
  }

  public int getDiscoveryPortOrZero() {
    return discoveryPort.orElse(0);
  }
```